chain/client/src/view_client_actor.rs

```diff
@@ -453,10 +446,38 @@ impl ViewClientActorInner {
                     block_height,
                     block_hash,
                 },
+                near_chain::near_chain_primitives::error::QueryError::NoGlobalContractCode {
+                    identifier,
+                    block_height,
+                    block_hash,
+                } => QueryError::NoGlobalContractCode { identifier, block_height, block_hash },
             }),
         }
     }
 
+    fn query_shard_uid(
+        &self,
+        request: &QueryRequest,
+        epoch_id: EpochId,
+    ) -> Result<ShardId, EpochError> {
+        match &request {
+            QueryRequest::ViewAccount { account_id, .. }
+            | QueryRequest::ViewState { account_id, .. }
+            | QueryRequest::ViewAccessKey { account_id, .. }
+            | QueryRequest::ViewAccessKeyList { account_id, .. }
+            | QueryRequest::CallFunction { account_id, .. }
+            | QueryRequest::ViewCode { account_id, .. } => {
+                account_id_to_shard_id(self.epoch_manager.as_ref(), account_id, &epoch_id)
+            }
+            QueryRequest::ViewGlobalContractCode { .. }
+            | QueryRequest::ViewGlobalContractCodeByAccountId { .. } => {
+                // for global contract queries we can use any shard_id, so just take the first one
+                let shard_ids = self.epoch_manager.shard_ids(&epoch_id)?;
+                Ok(*shard_ids.iter().next().expect("at least one shard should always exist"))
```

Comment:
cc @staffik Just musing - this will be fun when we move to single shard tracking RPC